### PR TITLE
Make sure KafkaConnect and KafkaConnectS2I can be scaled to zero

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- Bugfix

### Description

When KafkaConnect or KafkaConnectS2I are sclaed to 0 (`replicas=0`), they currently start failing. This PR updates the behaviour to:
* Make sure the Connect reocnciliations are passing in such situation
* Make sure Connector statuses are updated with an error in such situation

This PR additionally improves the behaviour when the connector reocnciliation failes for not being able to connect to the REST API also for other reasons and sets the appropriatte status in the connectors in that case as well.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java [hunk 1]
```diff
@@ -143,8 +144,15 @@ public KafkaConnectAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pf
                 .compose(i -> deploymentOperations.reconcile(namespace, connect.getName(), connect.generateDeployment(annotations, pfa.isOpenshift(), imagePullPolicy, imagePullSecrets)))
                 .compose(i -> deploymentOperations.scaleUp(namespace, connect.getName(), connect.getReplicas()))
                 .compose(i -> deploymentOperations.waitForObserved(namespace, connect.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> deploymentOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus))
+                .compose(i -> connect.getReplicas() > 0 ? deploymentOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs) : Future.succeededFuture())
+                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus)
+                        .recover(error -> {
+                            if (error instanceof ConnectTimeoutException && connect.getReplicas() == 0)   {
```
Comment (samuel-hawker) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: I might be wrong in saying this, but maybe we should tell `reconcileConnectors` to expect a `ConnectTimeoutException` if `replicas ==  0`? Then we don't have to capture the exception here.

(essentially move this logic into reconcileConnectors and change its behaviour.)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectS2IAssemblyOperator.java [hunk 2]
```diff
@@ -152,8 +153,15 @@ public KafkaConnectS2IAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability
                 .compose(i -> buildConfigOperations.reconcile(namespace, KafkaConnectS2IResources.buildConfigName(connect.getCluster()), connect.generateBuildConfig()))
                 .compose(i -> deploymentConfigOperations.scaleUp(namespace, connect.getName(), connect.getReplicas()))
                 .compose(i -> deploymentConfigOperations.waitForObserved(namespace, connect.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> deploymentConfigOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> reconcileConnectors(reconciliation, kafkaConnectS2I, kafkaConnectS2Istatus))
+                .compose(i -> connect.getReplicas() > 0 ? deploymentConfigOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs) : Future.succeededFuture())
+                .compose(i -> reconcileConnectors(reconciliation, kafkaConnectS2I, kafkaConnectS2Istatus)
+                        .recover(error -> {
+                            if (error instanceof ConnectTimeoutException && connect.getReplicas() == 0)   {
```
Comment (samuel-hawker) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectS2IAssemblyOperator.java:159]: ditto prev comment

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java (hunk 1 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: I had it that way originally. But it is not completely easy, because the connector reconciliation is triggered form two sources:
* `KafkaConnectAssemblyOperator` and `KafkaConnectS2IAssemblyOperator` which can provide the replicas.
* The KafkaConnector watch which doesn't have them. So it creates inconsistent results based on where it is triggered from and it seemed to be better this way.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ConnectorMockTest.java [hunk 3]
```diff
@@ -938,4 +939,133 @@ public void testConnectorPauseResume() {
...
+        when(api.getConnector(any(), anyInt(), any())).thenReturn(Future.failedFuture(new ConnectTimeoutException("connection timed out")));
+
+        Crds.kafkaConnectOperation(client).inNamespace(NAMESPACE).withName(connectName).edit()
+                .editSpec()
+                    .withReplicas(0)
+                .endSpec()
+                .done();
+
+        waitForConnectReady(connectName);
+        waitForConnectorNotReady(connectorName, "ConnectTimeoutException", "connection timed out");
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ConnectorMockTest.java:1003]: Should the status message include what the connection was being made _to_? "Connection to Connect REST API timed out", or "Connection to foo.bar:8083 timed out", or something like that?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java (hunk 1 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: Won't this error handling result in the `Ready` status on the `KafkaConnect`? It's ready in the sense that there wasn't an error in the reconciliation, but not in the sense that the cluster is 'ready' for REST requests, is that what we want?

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: It does, but that is the point of the whole issue. To stop the reocnciliation failing when scaled to 0. There will be no REST API, but the resource will be ready in the sense that it will run as requested from the user.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ConnectorMockTest.java (hunk 3 above)
Comment (scholzj) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ConnectorMockTest.java:1003]: Thsi is really just the test exception based on what the mocked APi throws. The atual exception would be much longer and will include the adress.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java (hunk 1 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: Fair point. But that means we're treating `NotReady` has having the semantics of "failed", rather than simply the cluster not being ready. Maybe we need a different condition entirely? I do feel that we need some way to distinguish a scaled-to-zero-like-you-asked kind of Ready from REST-interface-should-work kind of Ready, because people will want to automate on top of this status and so it needs to be unambiguous.

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: So what status would you set in such case?

Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: I'm not sure, are there any precedents in Kube with scale to zero (in `Deployment`, `ReplicaSet` or `StatefulSet`)? If Kube uses `Ready` in this case, maybe it would be enough to omit the REST connection information when scaled to zero.

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: Only Pod has `Ready` conditionm but there is no pod when it has 0 replicas. In deployment, the conditions do not show anything different from when they have replicas. They have for example the `Available` condition set to `True`.

Comment (samuel-hawker) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: I think it would make sense to me that we use a similar mechanism to what deployments and statefulsets use and say `Available` false and give a reason like: `NoReplicas` / `InvalidReplicas`, e.g. the condition would look something like:
```  
 - lastTransitionTime: "2020-04-29T09:26:39Z"
    lastUpdateTime: "2020-04-29T09:26:39Z"
    message: Scaling KafkaConnect to 0 is not supported.
    reason: InvalidReplicas
    status: "False"
    type: Available

```

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: They actually don't seem to do that. A deployment scaled to 0 has just this:

```
  status:
    conditions:
    - lastTransitionTime: "2020-04-29T08:15:31Z"
      lastUpdateTime: "2020-04-29T08:34:48Z"
      message: ReplicaSet "strimzi-cluster-operator-7bcd647d8d" has successfully progressed.
      reason: NewReplicaSetAvailable
      status: "True"
      type: Progressing
    - lastTransitionTime: "2020-04-29T11:18:52Z"
      lastUpdateTime: "2020-04-29T11:18:52Z"
      message: Deployment has minimum availability.
      reason: MinimumReplicasAvailable
      status: "True"
      type: Available
    observedGeneration: 3
```

So it actually says it is available. If we would go for it (regardless what Deployment does), the wierd thing would be whether it make sense to say Available / False when we do not have available / True when everything is working.

Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: Thinking more I don't think we should try to solve this using conditions, because we'll be forced into have two and that makes life harder at the point of use because to test for readiness you would need to check for both `Ready` and `Available` (or Ready and not Unavailable, or whatever). replicas=0 is a bit of an edge case which most people shouldn't have to deal with. 

So on reflection when replicas=0 I guess it would be OK to say `Ready=true`, but to not set the `status.url` or `status.connectorPlugins`. I think that would be enough of a signal for people wanting to use replicas=0.

Comment (samuel-hawker) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: Makes sense to me, my original thinking was that we would effectively override the deployments state of availability (that's why my suggestion differs from the deployment example Jakub linked) - but you are right querying two separate status conditions seems like extra complexity...

Tom's suggestion makes sense to me, though I do wonder if a user would be confused why the url/connectorPlugins fields are missing. Perhaps we follow the pattern that @dalelane introduced in #2663 and print an additional condition as a warning so the user has a little more to go on?

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: > Tom's suggestion makes sense to me, though I do wonder if a user would be confused why the url/connectorPlugins fields are missing. Perhaps we follow the pattern that @dalelane introduced in #2663 and print an additional condition as a warning so the user has a little more to go on?

I can do that if you want. Although the scaling to 0 is from my point of view always intentional and not done by mistake or not understanding how things work.

Comment (samuel-hawker) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: Very true, I agree it seems very unlikely a user would scale down to 0 and also expect a service to still exist. But i do still think an additional entry to conditions clarifying that ita ready but only in the trivial sense of that all containers (that is zero of them) are ready as there is nothing to do.

Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: What's the proposal exactly:

```yaml
type: "Available"
status: "False"
reason: "ScaledToZero"
```

Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: I'm not keen on calling it a warning. They asked for 0, and we gave them 0. There's no particular reason to treat that differently from other numbers of replicas. They asked for 3, but maybe they meant 5?

I think the warnings we added in #2663 were useful because they highlighted very definite cases where people could break their clusters/lose data. That's not the case here.

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: I'm not sure I like the _Available: False_ when we have no _Available: True_. But if that is fine for @samuel-hawker I can implement it.

Comment (samuel-hawker) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:150]: I agree that we should forego the Available condition, I wish I had made that a bit more clear further up the thread.
I agree that if a user asks for 0 replicas, we give them 0 replicas, I suggested a Warning only because a user in theory could wonder why the `status.url` or `status.connectorPlugins` fields are empty or missing.

Something like:
```
type: "Warning"
reason: "Zero replicas so no provided URL"
```
But maybe there is little value in this - what do you think @tombentley ?

**Comment by scholzj** [comment]
@samuel-hawker @tombentley I think I addressed the comments and applied the output of the long discussion here. Could you please have another look?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 4]
```diff
@@ -208,6 +209,11 @@ public void eventReceived(Action action, KafkaConnector kafkaConnector) {
                                                 log.info("{} {} in namespace {} was {}, but Connect cluster {} does not exist", connectorKind, connectorName, connectorNamespace, action, connectName);
                                                 updateStatus(noConnectCluster(connectNamespace, connectName), kafkaConnector, connectOperator.connectorOperator);
                                                 return Future.succeededFuture();
+                                            } else if ((connect != null && connect.getSpec() != null && connect.getSpec().getReplicas() == 0)
+                                                    || (connect != null && connect.getSpec() != null && connect.getSpec().getReplicas() == 0))    {
```
Comment (samuel-hawker) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:213]: Isn't this the same check twice?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java [hunk 5]
```diff
@@ -143,11 +143,14 @@ public KafkaConnectAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pf
...
                 .compose(i -> deploymentOperations.waitForObserved(namespace, connect.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> deploymentOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus))
+                .compose(i -> connect.getReplicas() > 0 ? deploymentOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs) : Future.succeededFuture())
+                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus, connect.getReplicas() == 0))
                 .setHandler(reconciliationResult -> {
                     StatusUtils.setStatusConditionAndObservedGeneration(kafkaConnect, kafkaConnectStatus, reconciliationResult);
-                    kafkaConnectStatus.setUrl(KafkaConnectResources.url(connect.getCluster(), namespace, KafkaConnectCluster.REST_API_PORT));
+
+                    if (connect.getReplicas() > 0) {
```
Comment (samuel-hawker) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:151]: Since we are now computing 
connect.getReplicas() > 0
connect.getReplicas() == 0
and > 0 again here is it worth setting a boolean before the compose something like:
```
connectHasZeroReplicas = connect.getReplicas() == 0
```
And then re-using this value for logic gating? I think it makes sense from the reconcile loop to have as little logic gating as possible and pre-declaring it on a one-time computed boolean might make the intent a little more clearer.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectS2IAssemblyOperator.java [hunk 6]
```diff
@@ -152,12 +152,15 @@ public KafkaConnectS2IAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability
...
-                .compose(i -> deploymentConfigOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> reconcileConnectors(reconciliation, kafkaConnectS2I, kafkaConnectS2Istatus))
+                .compose(i -> connect.getReplicas() > 0 ? deploymentConfigOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs) : Future.succeededFuture())
+                .compose(i -> reconcileConnectors(reconciliation, kafkaConnectS2I, kafkaConnectS2Istatus, connect.getReplicas() == 0))
                 .setHandler(reconciliationResult -> {
                     StatusUtils.setStatusConditionAndObservedGeneration(kafkaConnectS2I, kafkaConnectS2Istatus, reconciliationResult);
-                    kafkaConnectS2Istatus.setUrl(KafkaConnectS2IResources.url(connect.getCluster(), namespace, KafkaConnectS2ICluster.REST_API_PORT));
-                    kafkaConnectS2Istatus.setBuildConfigName(KafkaConnectS2IResources.buildConfigName(connect.getCluster()));
+
+                    if (connect.getReplicas() > 0) {
```
Comment (samuel-hawker) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectS2IAssemblyOperator.java:160]: ditto

**Comment by scholzj** [comment]
@samuel-hawker All good points. Thanks for raising them. They should be fixed now.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 7]
```diff
@@ -208,6 +209,11 @@ public void eventReceived(Action action, KafkaConnector kafkaConnector) {
...
                                                 updateStatus(noConnectCluster(connectNamespace, connectName), kafkaConnector, connectOperator.connectorOperator);
                                                 return Future.succeededFuture();
+                                            } else if ((connect != null && connect.getSpec() != null && connect.getSpec().getReplicas() == 0)
+                                                    || (connectS2i != null && connectS2i.getSpec() != null && connectS2i.getSpec().getReplicas() == 0))    {
+                                                log.info("{} {} in namespace {} was {}, but Connect cluster {} has 0 replicas", connectorKind, connectorName, connectorNamespace, action, connectName);
+                                                updateStatus(zeroReplicas(connectNamespace, connectName), kafkaConnector, connectOperator.connectorOperator);
+                                                return Future.succeededFuture();
                                             } else if (connect != null && isOlderOrAlone(connect.getMetadata().getCreationTimestamp(), connectS2i)) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:217]: What about the case where both connect and connectS2i are non null, but connectS2i is scaled to 0? We'd say there were zero replicas via the first branch, but we ought to proceed via the 2nd branch, right?

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java]: Good point. I pushed it one block below to make sure it is reflected which Connect takes the priority. Thanks
